File: main/dijkstrashortreach/dijkstrashortreach-fast.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;
import java.util.StringTokenizer;

/** A modified binary minheap, suitable for Prim's or Dijkstra's algorithms.
 *  Traditionally one adds contains() and decrease() methods. I have done this
 *  but also omitted the push() method, as those algorithms don't need it. */
final class PrimHeap {
    static final class KeyValuePair {
        KeyValuePair(final int key, final int value) {
            this.key = key;
            this.value = value;
        }

        final int key;
        int value;
    }

    /** Constructs a new heap with sequential keys in [first, last].
     *  Each key initially maps to the max value of an int (for "infinity"). */
    PrimHeap(int first, final int last) {
        if (first > last)
            throw new IllegalArgumentException("size must be nonnegative");

        final int len = last - first + 1;
        _table = new HashMap<>(len);
        _elems = new ArrayList<>(len);

        for (int i = 0; i != len; ++i) {
            _table.put(first, i);
            _elems.add(new KeyValuePair(first++, Integer.MAX_VALUE));
        }
    }

    boolean isEmpty() { return _elems.isEmpty(); }

    int size() { return _elems.size(); }

    boolean contains(final int key) { return _table.containsKey(key); }

    KeyValuePair pop() {
        final KeyValuePair ret = _elems.get(0);
        final int last = size() - 1;

        if (last == 0) {
            _table.clear();
            _elems.clear();
        } else {
            _table.remove(ret.key);
            _elems.set(0, _elems.remove(last));
            siftDown(0);
        }

        return ret;
    }

    void decrease(final int key, final int value) {
        final int child = _table.get(key);
        final KeyValuePair entry = _elems.get(child);

        if (value < entry.value) {
            entry.value = value;
            siftUp(child);
        }
    }

    private static final int NO_CHILD = -1;

    private void siftUp(int child) {
        while (child != 0) {
            final int parent = (child - 1) / 2;
            if (orderOK(parent, child)) break;

            swap(child, parent);
            child = parent;
        }

        record(child);
    }

    private void siftDown(int parent) {
        for (; ; ) {
            final int child = pickChild(parent);
            if (child == NO_CHILD || orderOK(parent, child)) break;

            swap(parent, child);
            parent = child;
        }

        record(parent);
    }

    private int pickChild(final int parent) {
        final int left = parent * 2 + 1;
        if (left >= size()) return NO_CHILD;

        final int right = left + 1;
        return right == size() || orderOK(left, right) ? left : right;
    }

    private void swap(final int src, final int dest) {
        final KeyValuePair tmp = _elems.get(dest);
        _elems.set(dest, _elems.get(src));
        _elems.set(src, tmp);

        record(src);
    }

    private void record(final int src) {
        _table.put(_elems.get(src).key, src);
    }

    private boolean orderOK(final int parent, final int child) {
        return _elems.get(parent).value <= _elems.get(child).value;
    }

    private final Map<Integer, Integer> _table; // key -> index
    private final List<KeyValuePair> _elems; // index -> (key, value)
}

/** A pair of vertices representing an undirected edge of unspecified weight.
 *  This type is effectively an implementation detail for the Solution class. */
final class Endpoints {
    Endpoints(final int u, final int v) {
        if (u > v) {
            this.u = v;
            this.v = u;
        } else {
            this.u = u;
            this.v = v;
        }
    }

    boolean equals(final Endpoints other) {
        return u == other.u && v == other.v;
    }

    @Override
    public boolean equals(final Object other) {
        // Endpoints is effectively an implementation detail. Comparing an
        // object of this type to an object of another type or to null would
        // always be a bug, and it is better for it to throw than return false.
        return equals((Endpoints)other);
    }

    @Override
    public int hashCode() {
        return (17 * 31 + u) * 31 + v;
    }

    final int u, v;
}

/** An element of a row of an adjacency list of a weighted undirected graph. */
final class Neighbor {
    Neighbor(final int dest, final int wt) {
        if (wt < 0)
            throw new IllegalArgumentException("negative weights not allowed");

        this.dest = dest;
        this.wt = wt;
    }

    final int dest, wt;
}

final class Solution {
    public static void main(final String[] args) throws IOException {
        try (final InputStreamReader isr = new InputStreamReader(System.in);
             final BufferedReader br = new BufferedReader(isr)) {
            for (int t = getVal(br); t > 0; --t) {
                final List<List<Neighbor>> adj = readGraph(br);
                final int start = getVal(br);

                final int[] costs = computeShortestReaches(adj, start, -1);

                final StringJoiner sj = new StringJoiner(" ");
                for (int i = MIN_VERTEX; i != costs.length; ++i)
                    if (i != start) sj.add(Integer.toString(costs[i]));

                System.out.println(sj);
            }
        }
    }

    private static final int MIN_VERTEX = 1; // the lowest-numbered vertex used

    /** Reads a graph as [order, size], [u1, v1, wt1], [u2, v2, wt2], ...
     * @param br    The BufferedReader used to read input.
     * @return The graph as an adjacency list. */
    private static List<List<Neighbor>> readGraph(final BufferedReader br)
            throws IOException {
        final StringTokenizer st = new StringTokenizer(br.readLine());
        final int vertexCount = getVal(st), edgeCount = getVal(st);

        final List<List<Neighbor>> adj = getEmptyAdjacencyList(vertexCount);

        readSimpleGraphEdges(br, edgeCount).forEach((uv, weight) -> {
            final int u = uv.u, v = uv.v, wt = weight;
            adj.get(u).add(new Neighbor(v, wt));
            adj.get(v).add(new Neighbor(u, wt));
        });

        return adj;
    }

    /** Creates an empty adjacency list: one initially empty row per vertex.
     * @param vertexCount   The number of vertices, starting from MIN_VERTEX.
     * @return An adjacency list of MIN_VERTEX + vertexCount empty rows. */
    private static List<List<Neighbor>> getEmptyAdjacencyList(int vertexCount) {
        if (vertexCount < 0)
            throw new IllegalArgumentException("negatively many vertices?");

        final int size = MIN_VERTEX + vertexCount;

        final List<List<Neighbor>> adj = new ArrayList<>(size);
        for (int n = MIN_VERTEX; n != 0; --n) adj.add(Collections.emptyList());
        while (vertexCount-- != 0) adj.add(new ArrayList<>());

        assert adj.size() == size;
        return adj;
    }

    /** Reads a graph as a compact adjacency matrix consisting of a map from
     *  endpoint pairs represented by Endpoints objects to Integer weights.
     * @param br        The BufferedReader used to read input.
     * @param edgeCount The "size" of the graph -- the number of edge to read.
     * @return  A map, edges, where edges.get(new Endpoints(u, v)) is the weight
     *          associated with the undirected edge (u, v). */
    private static Map<Endpoints, Integer> readSimpleGraphEdges(
            final BufferedReader br, int edgeCount) throws IOException {
        if (edgeCount < 0)
            throw new IllegalArgumentException("negatively many edges?");

        final Map<Endpoints, Integer> edges = new HashMap<>();

        while (edgeCount-- != 0) {
            final StringTokenizer st = new StringTokenizer(br.readLine());
            final int u = getVal(st), v = getVal(st), wt = getVal(st);
            if (u != v) edges.merge(new Endpoints(u, v), wt, Math::min);
        }

        return edges;
    }

    private static int getVal(final BufferedReader br) throws IOException {
        return Integer.parseInt(br.readLine());
    }

    private static int getVal(final StringTokenizer st) {
        return Integer.parseInt(st.nextToken());
    }

    /** Calculates all the shortest reaches in a graph from a given start
     * vertex, using Dijkstra's algorithm.
     * @param adj               An adjacency list for the graph.
     * @param start             The vertex to measure shortest paths from.
     * @param unreachableCost   The "cost" to indicate an unreachable vertex.
     * @return A costs array, where costs[i] is the cost of the shortest path
     *         from the starting vertex to vertex i, or unreachableCost. */
    private static int[] computeShortestReaches(final List<List<Neighbor>> adj,
                                                final int start,
                                                final int unreachableCost)
    {
        final int[] costs = new int[adj.size()];
        Arrays.fill(costs, unreachableCost);
        costs[start] = 0;

        final PrimHeap costHeap = new PrimHeap(MIN_VERTEX, adj.size() - 1);
        costHeap.decrease(start, 0);

        while (!costHeap.isEmpty()) {
            final PrimHeap.KeyValuePair entry = costHeap.pop();

            final int wt = entry.value;
            if (wt == Integer.MAX_VALUE) break;

            final int src = entry.key;
            costs[src] = entry.value;

            for (final Neighbor neighbor : adj.get(src)) {
                final int dest = neighbor.dest;
                if (costHeap.contains(dest))
                    costHeap.decrease(dest, wt + neighbor.wt);
            }
        }

        return costs;
    }

    private Solution() { throw new AssertionError(); }
}

File: main/dijkstrashortreach/dijkstrashortreach-fast.cs
#define DEBUG

using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Linq;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

/// <summary>
/// A modified binary minheap, suitable for Prim's and Dijkstra's algorithms.
/// </summary>
internal sealed class PrimHeap {
    internal PrimHeap(int minKey, int maxKey)
    {
        CheckConstructorArguments(minKey, maxKey);

        // Create and populate the [key -> index] map.
        _map = new int[maxKey + 1];
        for (var key = 0; key != minKey; ++key) _map[key] = None;
        var count = 0;
        for (var key = minKey; key != _map.Length; ++key) _map[key] = count++;
        Contract.Assert(count == maxKey - minKey + 1);

        // Create and populate the [index -> (key, value)] list.
        _entries = new List<KeyValuePair<int, int>>(count);
        while (count-- != 0)
            _entries.Add(new KeyValuePair<int, int>(minKey++, int.MaxValue));
        Contract.Assert(minKey == maxKey + 1);
    }

    internal int Count => _entries.Count;

    internal KeyValuePair<int, int> Pop()
    {
        var ret = _entries[0];
        _map[ret.Key] = None;

        var last = Count - 1;

        if (last == 0) {
            _entries.Clear();
        } else {
            _entries[0] = _entries[last];
            _entries.RemoveAt(last);
            SiftDown(0);
        }

        return ret;
    }

    internal void Decrease(int key, int value)
    {
        var child = _map[key];

        if (child != None && value < _entries[child].Value) {
            _entries[child] = new KeyValuePair<int, int>(key, value);
            SiftUp(child);
        }
    }

    private const int None = -1;

    private static void CheckConstructorArguments(int minKey, int maxKey)
    {
        if (minKey < 0) {
            throw new ArgumentOutOfRangeException(
                    paramName: nameof(minKey),
                    message: $"{nameof(minKey)} can't be negative");
        }

        if (minKey > maxKey) {
            throw new ArgumentException(
                    paramName: nameof(minKey),
                    message: $"{nameof(minKey)} can't exceed {nameof(maxKey)}"
                );
        }
    }

    private void SiftUp(int child)
    {
        while (child != 0) {
            var parent = (child - 1) / 2;
            if (OrderOK(parent, child)) break;

            Swap(child, parent);
            child = parent;
        }

        Record(child);
    }

    private void SiftDown(int parent)
    {
        for (; ; ) {
            var child = PickChild(parent);
            if (child == None || OrderOK(parent, child)) break;

            Swap(parent, child);
            parent = child;
        }

        Record(parent);
    }

    private int PickChild(int parent)
    {
        var left = parent * 2 + 1;
        if (left >= Count) return None;

        var right = left + 1;
        return right == Count || OrderOK(left, right) ? left : right;
    }

    private void Swap(int src, int dest)
    {
        var tmp = _entries[dest];
        _entries[dest] = _entries[src];
        _entries[src] = tmp;

        Record(src);
    }

    private void Record(int src)
        => _map[_entries[src].Key] = src;

    private bool OrderOK(int parent, int child)
        => _entries[parent].Value <= _entries[child].Value;

    // key -> index
    private readonly int[] _map;

    // index -> (key, value)
    private readonly List<KeyValuePair<int, int>> _entries;
}

/// <summary>
/// A pair of vertices representing the endpoints of an undirected graph.
/// </summary>
internal struct Endpoints : IEquatable<Endpoints> {
    internal Endpoints(int u, int v)
    {
        if (u > v) {
            V = u;
            U = v;
        } else {
            U = u;
            V = v;
        }
    }

    public bool Equals(Endpoints other)
        => U == other.U && V == other.V;

    public override bool Equals(object other)
        => Equals((Endpoints)other);

    public override int GetHashCode()
        => (17 * 31 + U) * 31 + V;

    internal int U { get; }
    internal int V { get; }
}

internal struct Neighbor {
    internal Neighbor(int dest, int wt)
    {
        Dest = dest;
        Wt = wt;
    }

    internal int Dest { get; }
    internal int Wt { get; }
}

internal static class Solution {
    /// <summary>The lowest vertex number used.</summary>
    private const int FirstVertex =  1;

    /// <summary>
    /// The value to show as the "cost" to reach an unreachable vertex.
    /// </summary>
    private const int UnreachableCost = -1;

    private static void Main()
    {
        for (var t = ReadValue(); t > 0; --t) {
            var adj = ReadGraph();
            var start = ReadValue();

            var costs = adj.ComputeShortestReaches(start);

            var vertices = Enumerable.Range(FirstVertex,
                                            adj.Count - FirstVertex);

            Console.WriteLine(string.Join(" ", from dest in vertices
                                               where dest != start
                                               select costs[dest]));
        }
    }

    private static int ReadValue() => int.Parse(Console.ReadLine());

    private static int[] ReadRecord()
    {
        var line = Console.ReadLine();

        var tokens = line.Split((char[])null,
                                StringSplitOptions.RemoveEmptyEntries);

        return Array.ConvertAll(tokens, int.Parse);
    }

    /// <summary>
    /// Reads a weighted directed graph as an adjacency list.
    /// </summary>
    private static List<List<Neighbor>> ReadGraph()
    {
        var vertexCount_edgeCount = ReadRecord();
        var vertexCount = vertexCount_edgeCount[0];
        var edgeCount = vertexCount_edgeCount[1];

        var adj = GetEmptyAdjacencyList(vertexCount);

        foreach (var edge in ReadSimpleGraphEdges(edgeCount)) {
            var u = edge.Key.U;
            var v = edge.Key.V;
            var wt = edge.Value;

            adj[u].Add(new Neighbor(v, wt));
            adj[v].Add(new Neighbor(u, wt));
        }

        return adj;
    }

    /// <summary>Creates an adjacency list with all rows empty.</summary>
    private static List<List<Neighbor>> GetEmptyAdjacencyList(int vertexCount)
    {
        if (vertexCount < 0) {
            throw new ArgumentOutOfRangeException(
                paramName: nameof(vertexCount),
                message: "a graph can't have negatively many vertices");
        }

        var size = FirstVertex + vertexCount;

        var adj = new List<List<Neighbor>>(size);
        for (var n = FirstVertex; n != 0; --n) adj.Add(null);
        while (vertexCount-- != 0) adj.Add(new List<Neighbor>());

        Contract.Assert(adj.Count == size);
        return adj;
    }

    /// <summary>
    /// Reads a graph as a hash-based sparse adjacency matrix.
    /// </summary>
    private static Dictionary<Endpoints, int>
    ReadSimpleGraphEdges(int edgeCount)
    {
        if (edgeCount < 0) {
            throw new ArgumentOutOfRangeException(
                paramName: nameof(edgeCount),
                message: "can't read negatively many edges");
        }

        var edges = new Dictionary<Endpoints, int>();

        while (edgeCount-- != 0) {
            var u_v_wt = ReadRecord();
            var u = u_v_wt[0];
            var v = u_v_wt[1];

            if (u == v) continue;

            var uv = new Endpoints(u, v);
            var wt = u_v_wt[2];

            int oldWt;
            if (!edges.TryGetValue(uv, out oldWt) || wt < oldWt)
                edges[uv] = wt;
        }

        return edges;
    }

    /// <summary>
    /// Calculates the least cost paths from a given starting vertex to each
    /// vertex in a graph, using Dijkstra's algorithm.
    /// </summary>
    /// <param name="adj">The graph to traverse, as an adjacency list.</param>
    /// <param name="start">The vertex where the traversal begins.</param>
    /// <returns>
    /// An array, <c>costs</c>, where <c>costs[index]</c> is the least cost to
    /// traverse from the vertex <c>start</c> to the vertex <c>index</c>.
    /// </returns>
    private static int[] ComputeShortestReaches(this List<List<Neighbor>> adj,
                                                int start)
    {
        var costs = new int[adj.Count];
        for (var i = 0; i != costs.Length; ++i) costs[i] = UnreachableCost;

        var costHeap = new PrimHeap(FirstVertex, adj.Count - 1);
        costHeap.Decrease(start, 0);

        while (costHeap.Count != 0) {
            var ent = costHeap.Pop();
            if (ent.Value == int.MaxValue) break;

            costs[ent.Key] = ent.Value;

            foreach (var neighbor in adj[ent.Key])
                costHeap.Decrease(neighbor.Dest, ent.Value + neighbor.Wt);
        }

        return costs;
    }
}
